pkg/credentials/asymmetric_key_signer.go

```diff
@@ -0,0 +1,145 @@
+package credentials
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+)
+
+type AsymmetricKeySignerConfig struct {
+	PrivateKey  string
+	CredId      string
+	AppOrigin   string
+	CrossOrigin *bool
+	Algorithm   *crypto.Hash
+}
+
+type AsymmetricKeySigner struct {
+	*AsymmetricKeySignerConfig
+}
+
+func NewAsymmetricKeySigner(config *AsymmetricKeySignerConfig) *AsymmetricKeySigner {
+	return &AsymmetricKeySigner{
+		AsymmetricKeySignerConfig: config,
+	}
+}
+
+// func stringToHash(hash *string) (*crypto.Hash, error) {
+// 	if hash == nil {
+// 		return nil, nil
+// 	}
+
+// 	i := 1
+// 	var h crypto.Hash
+// 	for crypto.Hash(i).Available() {
+// 		h = crypto.Hash(i)
+// 		if strings.EqualFold(h.String(), strings.ToUpper(*hash)) {
+// 			return &h, nil
+// 		}
+// 		i++
+// 	}
+
+// 	return nil, fmt.Errorf("hashing algorithm %s is not recognized", *hash)
+// }
+
+// Sign signs the given challenge using the private key and the hashing algorithm specified in the Algorithm field.
+// If the Algorithm field is not set or invalid, it defaults to SHA256.
+func (signer *AsymmetricKeySigner) Sign(challenge string, allowCredentials *AllowCredentials) (*KeyAssertion, error) {
+	// Determine the hashing algorithm
+	hash := signer.Algorithm
+	if hash == nil {
+		c := crypto.SHA256
+		hash = &c
+	}
+
+	// Parse PEM encoded private key
+	block, _ := pem.Decode([]byte(signer.PrivateKey))
+	if block == nil {
+		return nil, errors.New("invalid PEM data")
```

Comment:
would the error from `Decode` call contain more context on the reason of failure? we are losing that context if it does.

Reply:
Actually, no context is given:

```
func Decode(data []byte) (p *Block, rest []byte) {

```

Reply:
got it, thanks.